pywbem_mock/_mainprovider.py

```diff
@@ -0,0 +1,4208 @@
+
+        Returns a list of the reference instance names. The returned list is
+        the original, not a copy so the user must copy them
+        """
+
+        instance_store = self.get_instance_store(namespace)
+        class_store = self.get_class_store(namespace)
+
+        if assoc_class:
+            self._validate_class_exists(namespace, assoc_class, "AssocClass")
+
+        if result_class:
+            self._validate_class_exists(namespace, result_class, "ResultClass")
+
+        result_classes = self._subclasses_lc(result_class, class_store)
+        assoc_classes = self._subclasses_lc(assoc_class, class_store)
+
+        inst_name.namespace = namespace
+        role = role.lower() if role else role
+        result_role = result_role.lower() if result_role else result_role
+
+        ref_paths = self._get_reference_instnames(namespace, inst_name,
+                                                  assoc_class, role)
+
+        # Get associated instance names
+        rtn_instpaths = set()
+        for ref_path in ref_paths:
+            inst = self._find_instance(ref_path, instance_store)
+            for prop in six.itervalues(inst.properties):
+                if prop.type == 'reference':
+                    if prop.value == inst_name:
+                        if assoc_class \
+                                and inst.classname.lower() not in assoc_classes:
+                            continue
+                        if role and prop.name.lower() != role:
+                            continue
+                    else:
+                        if result_class and (prop.value.classname.lower()
+                                             not in result_classes):
+                            continue
+                        if result_role and prop.name.lower() != result_role:
+                            continue
+                        rtn_instpaths.add(prop.value)
+        return rtn_instpaths
+
+    ####################################################################
+    #
+    #   ReferenceNames server request
+    #
+    ####################################################################
+
+    def ReferenceNames(self, namespace, ObjectName, ResultClass=None,
+                       Role=None):
+        """
+        Implements a WBEM server responder for
+        :meth:`~pywbem.WBEMConnection.ReferenceNames`
+
+        Retrieves the instance paths of the association instances that
+        reference a source instance, or the class paths of the association
+        classes that reference a source class from the CIM instances in the CIM
+        repository. The association instances defining the relationship between
+        ObjectName and result classes must exist in the CIM repository.
+
+        Parameters:
+
+          namespace (:term:`string`):
+            Name of the namespace in which the CIM instance will be created
+              (case independent).
```

Comment:
Copy & paste error from CreateInstance. Should use enumeration verbiage.

Reply:
DONE